src/create-template-orders.js

```diff
@@ -20,20 +33,263 @@ async function _fetchLastStartTimestamp() {
+    nextDeliveryDate.setDate(nextDeliveryDate.getDate() - cutoffDays)
+    if (nextDeliveryDate.toISOString() < checkoutOrder.createdAt) {
+      nextDeliveryDateISOString = nextDeliveryDate.toISOString()
+      nextDeliveryCronDate = cronExpression.next()
+    }
+  }
+  const reminderDays = lineItem.custom.fields.reminderDays
+  let nextReminderDateISOString
+  if (reminderDays) {
+    nextDeliveryCronDate.setDate(nextDeliveryCronDate.getDate() - reminderDays)
+    nextReminderDateISOString = nextDeliveryCronDate.toISOString()
+  }
+  const templateOrder = {
+    orderNumber:
+      lineItem.custom.fields.subscriptionKey +
+      (quantityIncrement ? `-${quantityIncrement}` : ''),
+    customerEmail: checkoutOrder.customerEmail,
+    totalPrice: lineItem.totalPrice,
+    lineItems: [lineItem],
+    custom: {
+      type: {
+        typeId: 'type',
+        key: 'subscription-template-order',
+      },
+      fields: {
+        ...checkoutOrder.custom.fields,
+        nextDeliveryDate: nextDeliveryDateISOString,
+        schedule,
+        checkoutOrderRef: { typeId: 'order', id: checkoutOrder.id },
+        reminderDays,
+        nextReminderDate: nextReminderDateISOString,
+      },
+    },
+    inventoryMode: 'None',
+    state: {
+      typeId: 'state',
+      key: 'Active',
+    },
+  }
+  delete templateOrder.custom.fields.hasSubscription
+  return templateOrder
+}
+
+function _createPaymentDraft(payment) {
+  const paymentDraft = _.cloneDeep(payment)
+  delete paymentDraft.key
+  return paymentDraft
+}
+
+async function _buildFetchCheckoutOrdersUri() {
   let where =
     'custom(fields(hasSubscription=true)) AND custom(fields(isSubscriptionProcessed is not defined))'
   const lastStartTimestamp = await _fetchLastStartTimestamp()
   if (lastStartTimestamp)
-    where = `createdAt > ${lastStartTimestamp} AND ${where}`
-  const {
-    body: { results },
-  } = await apiRoot.orders().get({ queryArgs: { where } }).execute()
-  return results
+    where = `createdAt > "${lastStartTimestamp.value}" AND ${where}`
+
+  const uri = ctpClient.builder.orders
+    .where(where)
+    .expand('paymentInfo.payments[*]')
+  return uri
 }
 
-async function createTemplateOrders() {
-  const unprocessedCheckoutOrders = await fetchCheckoutOrders()
+async function _updateLastStartTimestamp(startDate) {
+  await apiRoot
+    .customObjects()
+    .post({
+      body: {
+        container: LAST_START_TIMESTAMP_CUSTOM_OBJECT_CONTAINER,
+        key: LAST_START_TIMESTAMP_CUSTOM_OBJECT_KEY,
+        value: startDate.toISOString(),
+      },
+    })
+    .execute()
+}
+
+async function createTemplateOrders(startDate) {
+  const uri = await _buildFetchCheckoutOrdersUri()
+
+  const processCheckoutOrderPromises = []
+  await ctpClient.fetchBatches(uri, (results) => {
+    for (const order of results)
+      processCheckoutOrderPromises.push(_processCheckoutOrder(order))
+  })
+  await pMap(processCheckoutOrderPromises, (f) => f, { concurrency: 3 })
```

Comment:
I think most of the promises will be fullfilled until this line:290, because fetchBatches will be waiting 🤔 also I'm thinking that maybe fetching and waiting until all orders created might be better, wdyt ?

Reply:
I think it works like this - it will call `_processCheckoutOrder` and in this function it will go to the first await. Once it gets to the first await in the function `_processCheckoutOrder` it will stop and return a promise that will be put into `processCheckoutOrderPromises` array. At least that is how await works. So it won't be finished once it gets to `await pMap`: https://exploringjs.com/impatient-js/ch_async-functions.html#executing-async-functions-synchronous-start-asynchronous-settlement-advanced

What is your suggestion? Can you write a code snippet here?

Reply:
First of all this code hard to reason, for instance it keep states of the returned promises in array and awaiting them with pMap, calling async functions without awaiting etc, so to me it looks like it's mixed of callback, async/await and promises. 

I've not tested it, but what I know the await will pause the function and then resume it back when it's added to call stack after the event loop pulls it from the callback queue. On every iteration of the fetchBatches the function will add promises without awaiting so assuming it might not be concurrent as it's supposed to be, therefore I assume triggered async functions might be fulfilled already before the pMap (which is tend to be used to limit concurrency). Correct me if I'm wrong on my assumption.

I would add pMap to line 288 and await all orders fetched with paginated query is processed before calling another batch on process function for the remaining orders. 

For a pseudocode snippet I was thinking something like below, but callback style on the fetchBatches method also makes it hard to reason and I'm not sure the snippet below works paralllel with .process method on sdk.

````js
await ctpClient.fetchBatches(uri, async (orders) => {
    stats.processedCheckoutOrders += orders.length
    await pMap(orders,
        async (order) => { await _processCheckoutOrder(order) },
        { concurrency: 3 })
  })
````

Reply:
I checked my code using the test and breakpoints and it works fine. It has a bit better concurrency contro, but I guess it's hard to read, so I updated to your version, which is also good and easier to read: https://github.com/commercetools/commercetools-subscriptions/pull/5/commits/e26c83f59e73e5abd836dec9a4d58d9325bf91ab